fenapack/field_split.py

```diff
@@ -15,159 +15,168 @@
+        all subKSP objects.
         """
-        dolfin.PETScKrylovSolver.set_operators(self, A, P)
-        assert self._ksp.getOperators() == \
-          (dolfin.as_backend_type(A).mat(), dolfin.as_backend_type(P).mat())
-
-        # Set up KSP
-        self._set_from_parameters() # update global option database
-        self._ksp.setFromOptions()
-        self._ksp.setUp() # NOTE: this includes operations within 'PCSetUp'
-
-        # Get subKSP and subPC objects
-        ksp0, ksp1 = self._ksp.getPC().getFieldSplitSubKSP()
-        pc0, pc1 = ksp0.getPC(), ksp1.getPC()
-
-        # Check if python context has been set up to define approximation of
-        # 11-block inverse. If so, use **schur_approx to set up this context.
-        if self._OptDB_11.hasName("pc_python_type"):
-            ctx = pc1.getPythonContext()
-            ctx.set_operators(self._is0, self._is1, A, P, **schur_approx)
-
-        # Set up each subPC explicitly before calling 'self.solve'. In such
-        # a case, the time needed for setup is not included in timings under
-        # "PETSc Krylov Solver".
-        timer = dolfin.Timer("FENaPack: set up subPC object pc0")
-        dolfin.log(dolfin.PROGRESS, "Preparing for the use of pc0 (calling PCSetUp).")
-        pc0.setUp()
-        timer.stop()
-        timer = dolfin.Timer("FENaPack: set up subPC object pc1")
-        dolfin.log(dolfin.PROGRESS, "Preparing for the use of pc1 (calling PCSetUp).")
-        pc1.setUp()
-        timer.stop()
-
-
-    def _set_from_parameters(self):
-        """Set up extra parameters added to parent class."""
-        # Get access to global option database
-        OptDB = PETSc.Options(self._ksp.getOptionsPrefix())
-        #OptDB["help"] = True
-
-        # Add extra solver parameters to the global option database
-        prm = self.parameters["preconditioner"]
-        OptDB["ksp_pc_side"] = \
-          prm["side"]
-        OptDB["pc_fieldsplit_type"] = \
-          prm["fieldsplit"]["type"]
-        OptDB["pc_fieldsplit_schur_fact_type"] = \
-          prm["fieldsplit"]["schur"]["fact_type"]
-        OptDB["pc_fieldsplit_schur_precondition"] = \
-          prm["fieldsplit"]["schur"]["precondition"]
-
-        # FIXME: Sort out a way how to deal with parameters;
-        #        if one uses directly ksp() object, then DOLFIN
-        #        parameters are not used, that's why this workaround;
-        #        maybe rather use only petsc4py api and don't mess up
-        #        with any parameters
-        OptDB["ksp_gmres_restart"] = \
-          self.parameters["gmres"]["restart"]
+        self._ksp.init_pcd(pcd_problem, pcd_pc_class=pcd_pc_class)
+
+
+    def ksp(self):
+        return self._ksp
+
+
+    ksp.__doc__ = PETScKrylovSolver.ksp.__doc__
+
+
+    def set_options_prefix(self, prefix):
+        self._ksp.setOptionsPrefix(prefix)
+
+
+    set_options_prefix.__doc__ = PETScKrylovSolver.set_options_prefix.__doc__
```

Comment:
Yes, this is ugly!

We don't use `KSPPYTHON` (for implementing `PCDKSP`) here which might have the right directors to its methods implemented in Python, because we want to use just GMRES implemented directly in PETSc and avoid implementing it by ourself.

We overload some methods of `petsc4py.PETSc.KSP` on Python side but C/C++ layer does not see that (because this was not intended). So call to `PETScKrylovSolver.set_options_prefix` would not call in-Python-overloaded methods but the original one implemented in C. And `PETScKrylovSolver.ksp()` would return `petsc4py.PETSc.KSP` object not having the Python overloads.

This is a hack and it would be better to sort this out but it might not be easy without changes to DOLFIN and/or petsc4py.